packages/flutter_tools/lib/src/resident_runner.dart

```diff
@@ -1502,6 +1502,68 @@ abstract class ResidentRunner extends ResidentHandlers {
     appFinished();
   }
 
+  @protected
+  Future<void> performPreHotRestart() async {
+    final Timer slowCallbackTimer = Timer(const Duration(seconds: 5), () async {
+      globals.printError('preHotRestartCallbacks are taking longer than expected...');
+    });
```

Comment:
We should probably have a feature that allows the user to hit `R` again to force the hot restart without waiting for the preHotRestartCallbacks if they tap it while one is already happening, in case one of the callbacks never returns.

Reply:
I've been exploring this behavior in a separate branch if you want to take a look at it https://github.com/davidmartos96/flutter/tree/abort_previous_pre_hot_restart

Do we want this behavior in this PR or explore it in a new one after this is merged? The terminal handler does not support running 2 commands at the same time. So pressing R while a hot restart is running is a no-op. I had to patch it for it to work that way in the terminal code, but the VSCode extension (and probably the IntelliJ as well) also don't allow pressing hot restart while one is running. 
From the tests I did it seems to correctly abort the pre hot restart callbacks, but I may have missed something by allowing to trigger a new hot restart while one is active.

Reply:
I would say that should be tracked separately in a feature request. It's unclear to me if we should do that, and if so it should probably be supported in the IDEs too.